update docs for new campaign flow

sqs opened the pull request:
I am pre-writing the docs so we have a concise description of the new flow from the user's POV.

**Preview as rendered docs at:**

- https://docs.sourcegraph.com/@campaigns-new-flow/user/campaigns
- https://docs.sourcegraph.com/@campaigns-new-flow/user/campaigns/hello_world_campaign
- https://docs.sourcegraph.com/@campaigns-new-flow/dev/campaigns_design

**Figma mock (slightly behind the docs pages above):** https://www.figma.com/file/kRf5i7xf1p4zs3OyX0DUVp/Campaigns


---

Docs that led to this change: [RFC 157](https://docs.google.com/document/d/1-Ed6s3yiB-pmL7Up8H6nMlUhgmLn5BVu34Zpvf-_YNo/edit#) and [Campaigns CLI and naming/concepts idea](https://docs.google.com/document/d/1zsN7NND7Lyc7Cm2KNdPYLwYs6pV7fvuPwLunV_u_tnY/edit#heading=h.by0rs1hib7mp). The proposals and motivations from those docs will be fully incorporated into this PR's `.md` doc changes.

sqs commented on doc/user/campaigns/index.md:
```diff
@@ -1,90 +1,121 @@
 # Campaigns
```
This doc page is much easier to read at http://docs.sourcegraph.com/@campaigns-new-flow/user/campaigns than in this diff.

sqs commented on doc/user/campaigns/index.md:
```diff
@@ -1,90 +1,121 @@
 
-## What are campaigns?
+The changes made by campaigns usually fit into a few general categories:
 
```
Need to clarify that YOU (the user) choose what type of changes to make, it’s not something automatic or (yet) built in

sqs commented:
I started on the wireframes for the new flow, but I got stuck on how to handle the fact that both editing name/description or patches can trigger an update to all changesets. Need to work on this some more, hopefully tomorrow morning PDT.

sqs commented on doc/user/campaigns/managing_access.md:
```diff
@@ -0,0 +1,78 @@
+- Adding existing changesets to a campaign: You must have read access to the existing changesets' repository.
+- Closing or deleting a campaign: If you choose to also close associated changesets on the code host, you must have access to do so on the code host. If you do not have access to close a changeset on the code host, the changeset will remain in its current state. A person with repository permissions for the remaining changesets can view them and manually close them.
+
+Your repository permissions can change at any time. If you've already published a changeset to a repository that you no longer have access to, then you won't be able to view its details or update it in your campaign. The changeset on the code host will remain in its current state. A person with permissions to the changeset on the code host will need to manually manage or close it. When possible,  ...
```
Cover when you lose write access to a repository then update a campaign. Your patchset doesn't cover that repository, and so the repository's changeset is left untouched.

eseliger commented on doc/user/campaigns/hello_world_campaign.md:
````diff
@@ -0,0 +1,93 @@
+
+Save the following campaign spec as `hello-world.campaign.yaml`:
+
+```yaml
````
It just crossed my mind that it might even be useful to say

```
changesetTemplate:
  ...
  patch:
    generate:
      - run: blah
```

(essentially, making patch also a property of the template, instead of the top-level `steps`). That way, we could in the future allow to add existing patches for example using
```
patch:
  fromFile: awesome_file.patch
```

sqs replied:
When would you prefer to use existing patches vs. generating them?

(I think this suggestion would violate the proposed "Define a campaign in a file" and "Shareable and portable" principles at https://docs.sourcegraph.com/@campaigns-new-flow/dev/campaigns_design  because you would need those other files as well to compose the campaign. Those principles are obviously still WIP.)

sqs commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -893,6 +936,52 @@ type ChangesetConnection {
 
     # Pagination information.
     pageInfo: PageInfo!
+
```
Ignore the rest of the GraphQL diff below here. It was just for the UI mockup, it's not an important part of the proposed API for the new campaigns flow.

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,109 @@ type Mutation {
+
+    # TODO(sqs): do we really need create/apply/update? or can we combine them? I think we do, but we might be able to make them a single mutation with multiple params (although that might be more complex).
+    #
+    # Why is createCampaign needed? It would be weird if in the UI when you create a new empty campaign, if your campaign name is the same as one that exists, if it just took you to the existing one. In that case (given that the button is 'Create campaign'), it should show an error.
```
@eseliger's question here: do we really need to support creating a campaign in the UI? (Or could we just provide instructions for creating a campaign on the CLI). Erik is concerned that the workflow in which you first create a campaign in the UI and then have to copy the name to a campaign spec YAML file feels weird.

sqs replied:
The UI empty state would tell you how to create it in the CLI. I think this would be a huge improvement in usability for a small amount of additional work. It would let us give the precise YAML stub file with `name: ___` and the `src campaign apply -namespace ___`, which I think would make it easier for people who are starting out.

eseliger replied:
Hmm I think it's a bit clunky copying a file content, putting that into a file, modifying it and running a src command.
Maybe it could be easier to do using `src campaign template -name=BLAH -out campaign.yml` (so just a copyable one-liner)?

sqs replied:
I don't think it will be clunky. I'll include it in a mock and we can see.

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,109 @@ type Mutation {
+    #
+    # Why is applyCampaign needed? This is the most obvious/general one. We could make the CLI decide whether to use createCampaign or updateCampaign by first seeing if the campaign existed, but that seems like logic that should not need to be implemented by each API consumer.
+    #
+    # Why is updateCampaign needed? Renaming campaigns requires it. Also, it is good practice for most API consumers to use a stable/immutable ID for a campaign (kind of like how we use the github numeric repo ID in our own DB to refer to GitHub repos).
```
See above, we think that we might not need `updateCampaign` when all we need is `renameCampaign`

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,109 @@ type Mutation {
+    # TODO(sqs): do we need 1 previewXyzCampaign mutation for each? or can we combine these into 1?
+    previewCreateCampaign(namespace: ID!, spec: CampaignSpecInput!): CampaignUpdatePreview!
+    previewApplyCampaign(namespace: ID!, spec: CampaignSpecInput!, lastSeq: Int): CampaignUpdatePreview!
+    previewUpdateCampaign(campaign: ID!, spec: CampaignSpecInput!, lastSeq: Int): CampaignUpdatePreview!
```
Super interested in what `CampaignUpdatePreview` contains.

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,109 @@ type Mutation {
+    # campaign would sit around empty. That might even be the desired behavior, esp. since you
+    # wouldn't create a bunch of duplicates if you retried because the name would be the same each
+    # time.
+    uploadChangesetSpec(campaign: ID!, changesetSpec: String!): ID!
```
Few thoughts:

1. I think this mutation can be confusing to API clients, because it's not clear whether uploaded changeset specs are "live" immediately (meaning: the reconciler picks them up right away after uploading) or whether they need to be put into a `CampaignSpecInput` which is then used in an `apply`/`update`. Erik, for example, assumed that they would go live right away, meaning that updating/previewing would be difficult. I assumed that newly uploaded changesetSpecs are just stored somewhere in the DB and only picked up by the reconciler as soon as they're applied to a campaign.
2. if `changesetSpecs` are uploaded individually and not collected in a set/list/connection, updating a campaign would be a bit cumbersome for clients. Say `src-cli` produced 5000 patches and uploads them in batches of 1000s. It would need to remember all 5000 IDs to then be able to do the correctly `apply`/`update`.
3. This `uploadChangesetSpec` should, in any case, accept an array of strings and return an array of IDs.
4. When would "unused" changeset specs be deleted? (If the user just previews an update and then decides to not apply it, we'd have changesetspecs associated with a campaign, but basically unfindable)

sqs replied:
Thanks for these. I'll take another spin in a couple hours.

sqs replied:
Is (2) really a problem? Avoiding this would require making the API a lot more complex, as I see it. It would require us to make a new mutable API object like `ChangesetSpecList`, and 2 new mutations `appendToChangesetSpecList` and `sealChangesetSpecList` (because I think it's important to mark it as immutable when done, to avoid confusion).

sqs replied:
(3) You can do batching in GraphQL itself (i.e., in a single request, you can specify multiple of the same mutation) if you really need to. Do we need to add batching to the mutation's parameters as well? I think each changeset spec should be in its own HTTP request, anyway, because otherwise we'd be increasing the risk of having a too-large request. I doubt the performance hit is going to be noticeable.

sqs replied:
(1) I think we can handle this in documentation. Also, accepting only a single changeset spec (ie not having the mutation accept upload of multiple changeset specs in one call) would reduce confusion here since fewer people would even think that uploading a single changeset spec would take effect immediately.

sqs replied:
See updated proposed API on this branch.

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,109 @@ type Mutation {
+    # This mutation is for callers that intend to rename the campaign or otherwise need/want to use
+    # a stable, immutable ID to identify the campaign to update. Otherwise, callers can use the
+    # applyCampaign mutation.
+    updateCampaign(
```
Question that came up while @eseliger and I were talking about this on Zoom: are we going to allow updating of campaign fields in the UI?

sqs replied:
No, because we wouldn't be able to generate the ChangesetSpecs for the updated CampaignSpec in the UI. (For extensions like the find-replace campaign extension, yes, it could update campaign fields "in the UI", but that's not the general case.)

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,109 @@ type Mutation {
+
+    # Changeset specs that were previously uploaded using the uploadChangesetSpec mutation. This
+    # field's items are merged with those from the changesetSpecs field.
+    changesetSpecIDs: [ID!]
```
One overall thought I have when thinking about this new flow from an architectural/implementation perspective: I really don't like our trifecta of `Patch`/`ChangesetJob`/`Changeset` anymore and I'd love to get rid of at least one of these three tables.

Having three tables makes the update logic complex ("okay, the new patches are in repos A, B, C. The existing changesets have patches in B, C, so we need to rewire the changesetjobs for these existing changesets to point at new patches in A, B, C.") and is just hard to think about.

In an ideal world, we could add a `spec`/`changesetInput`/etc. column to `changesets`. In the update, we'd look at the new changeset specs, find/create the changesets that match the repo/branch of the spec, and update their `spec`. The reconciler/syncer/job-executor (doesn't matter) would then simply detect that `changeset.specs` columns have been updated and ensure that the state of the changeset on the code host matches the new desired state.

Now, one could rightfully ask "well, thorsten, if you don't like 3 tables, why'd you create them?" Yes, good point. But I think in the old, imperative way where a user would actively say "publish this, publish that" it made more sense to create a `ChangesetJob`/`Changeset` only when a diff (that has been reviewed) has been published.

With the new flow, however, I think it makes more sense to have `Changeset` be a highly mutable database entry that's always updated to ensure that its remote counterpart matches what's specified in its `spec`.

Why am I writing this down here? :) Just to add some more food for thought here, because I think this is relevant to how we design the "uploadChangesetSpec" part of the API. That — having changesetSpecs lying around that are not "ready to be reconciled" until applied to a campaign — makes me think that this is just another form of the `patches` table.

sqs commented on doc/user/campaigns/hello_world_campaign.md:
````diff
@@ -0,0 +1,93 @@
+Save the following campaign spec as `hello-world.campaign.yaml`:
+
+```yaml
+name: hello-world
````
Yeah, I would propose something like this in the JSON Schema (which can also be used to validate equivalent YAML):

```json
"name": {
      "description": "The name of the campaign, which is unique among all campaigns in the namespace. A campaign's name is case-preserving.",
      "pattern": "^[\\w.-]+$"
    },
```

sqs commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,109 @@ type Mutation {
+        # TODO(sqs): This is a way you can avoid race conditions when you update a campaign that had
+        # a newer state from what you expected, like lastID in our settings API. Need to add details
+        # here.
+        lastSeq: Int
```
I changed my mind and actually don't think this is needed.

sqs commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,109 @@ type Mutation {
+    # Update a campaign identified by ID (not name) from a campaign spec and locally computed
+    # changeset specs. The campaign is returned.
+    #
+    # This mutation is for callers that intend to rename the campaign or otherwise need/want to use
```
What do you think about this justification?

> Why is updateCampaign needed? Renaming campaigns requires it. Also, it is good practice for most API consumers to use a stable/immutable ID for a campaign (kind of like how we use the github numeric repo ID in our own DB to refer to GitHub repos).

sqs commented:
@eseliger @mrnugget See updated GraphQL API: https://github.com/sourcegraph/sourcegraph/pull/10921/files?file-filters%5B%5D=.graphql#diff-4529437f5564f5e65e957821a46ff9c5R372.

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,124 @@ type Mutation {
+    ): Campaign!
+
+    # Move a campaign to a different namespace, or rename it in the current namespace.
+    moveCampaign(campaign: ID!, newName: String, newNamespace: ID): Campaign!
```
I like this better than the `updateCampaign`

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,124 @@ type Mutation {
+        # TODO - this is kinda weird but seems useful to avoid someone accidentally clobbering
+        # changes on a campaign when someone else renamed/moved it without their knowledge (eg if
+        # they kept a browser tab open or reused a CLI history entry).
+        campaign: ID
```
I think that makes sense.

Question is: by setting `campaign` one can use `applyCampaign` to rename a campaign. Would we then need the `newName` param in `moveCampaign`?

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,124 @@ type Mutation {
+    ): Campaign!
+
+    # Delete a campaign. A deleted campaign is completely removed and can't be un-deleted. The
+    # campaign's changesets are kept as-is; to close them, use the closeCampaign mutation first.
```
I like this!

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,124 @@ type Mutation {
+    #
+    # The returned ChangesetSpec is immutable and expires after a certain period of time (if not
+    # used in a call to applyCampaign), which can be queried on ChangesetSpec.expiresAt.
+    uploadChangesetSpec(
```
The other mutation is called `createCampaignSpec` so I think we should be consistent here:

```suggestion
    createChangesetSpec(
```

mrnugget replied:
I like this mutation. It's what I had in mind in an earlier draft with `createDiff`.

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,124 @@ type Mutation {
+    # Compute the delta (difference) between a campaign's desired state (spec) and its actual state
+    # (status).
+    #
+    # NOTE: This actually belongs in Query (not Mutation), but I put it here for ease of review in
```
Why a query? It sounds like the `CampaignDelta` is persisted and can be addressed by ID. That makes me think it's a mutation and not a query.

sqs replied:
Yeah, you are right

sqs commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,124 @@ type Mutation {
+type CampaignSpec implements Node {
+    changesetSpecs: [ChangesetSpec!]!
+    expiresAt: DateTime
+}
```
I would add some fields like:

```
creator: User
createdAt: DateTime!
raw: String! # raw YAML
```

but not actually parse out the YAML and expose all those things as fields.

mrnugget replied:
That makes sense and, yeah, would not expose those fields and, in general, would rather have a schema for the spec and pass around strings in the API than to have a "strongly typed" API in which every field is exposed.

sqs commented:
@mrnugget:

> I'm worried that adding ChangesetSpecs as separate entities (addressable by ID) would not simplify things a lot on the backend. One idea I had was to have something like createChangeset(spec: ChangesetSpecInput): Changeset! but then have the reconciler only pick up the changeset as soon as it's attached to a campaign. The spec would be saved in a column on changeset. That would've been the changeset-equivalent to `createCampaign(spec: CampaignSpecInput: changesets: [ID!]): Campaign`. That would simplify things in the backend, I think. But feel free to ignore that.

(I renamed `{upload => create}ChangesetSpec`.)

I don't like `createChangeset(ChangesetSpecInput): Changeset` because it violates the model of "specs need to be applied before they are used". It would create a 3rd "phase of an object": "just an unapplied spec", "applied and has a current state", and the new "just an unapplied spec but uses the type (`Changeset`) of an applied thing with current state".

Anyway, if this helps, in the backend, you can still create a `changesets` row when `createChangesetSpec` is called. Its `spec` column will be set but its `status` (or whatever) column will not be set.

mrnugget commented:
> because it violates the model of "specs need to be applied before they are used".

Got it! That's a good way of thinking about it (and, in hindsight, the obvious way 😉 )

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,125 @@ type Mutation {
+
+# An immutable description of the desired state of a campaign.
+type CampaignSpec implements Node {
+    changesetSpecs: [ChangesetSpec!]!
```
```suggestion
    id: ID!
    changesetSpecs: [ChangesetSpec!]!
```

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,125 @@ type Mutation {
+
+# An immutable description of the desired state of a changeset in a campaign.
+type ChangesetSpec implements Node {
+    id: ID! # unguessable ID
```
@sqs I assume "unguessable ID" means that it's not just the base64 encoding of "ChangesetSpec:<databaseID>". Correct?

sqs replied:
yes, added more docstring here

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -32,66 +32,6 @@ scalar JSONCString
 
 # A mutation.
 type Mutation {
-    # Create a list of changesets in a specific repository on a code host (e.g.,
-    # pull requests on GitHub). If a changeset with the given input already
-    # exists, it is returned instead of a new entry being added to the database.
-    createChangesets(input: [CreateChangesetInput!]!): [ExternalChangeset!]!
-    # Add a list of changesets to a campaign. The campaign must not have a patchset.
-    addChangesetsToCampaign(campaign: ID!, changesets: [ID!]!): Campaign!
```
Do we want to remove these two in the first iteration of the new workflow?

If so, we need to specify how to list manual changesets in the campaign spec.

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,125 @@ type Mutation {
+        # The campaign spec as YAML (TODO also allow JSON for easier machine-creation or creation from JS web clients?).
+        campaignSpec: String!
+        # Changeset specs that were locally computed and then uploaded using createChangesetSpec.
+        changesetSpecs: [ID!]!
```
Should this also use `input: CampaignSpecInput!`, like `computeCampaignDelta` further down? (The `CampaignSpecInput` type is not defined)

sqs replied:
I removed CampaignSpecInput for simplicity. We will require a call to createCampaignSpec to use a campaign spec and won't support just passing in the YAML directly. This is simpler and actually lets us have URLs to preview "what will happen when I create/update this campaign" that use the campaign spec ID instead of making a new type like CampaignDelta that is persisted.

sqs commented:
@mrnugget @eseliger I updated the schema.graphql file. Main changes are (1) requiring you to always call createCampaignSpec to create a campaign spec, no more mutations that let you just upload the YAML and (2) made CampaignDelta a non-persisted type. Also added a lot more docstrings. PTAL.

How common operations are implemented:

- Creating a campaign: `src campaigns apply -f foo.campaign.yaml [-namespace alice|myorg] -preview` -> computes changesets locally, calls `createChangesetSpec` 1x per changeset, calls `createCampaignSpec` 1x, prints out preview URL `CampaignSpec.previewURL`, which is like `https://sourcegraph.example.com/users/alice/campaigns/new?spec=CAMPAIGN_SPEC_ID`
- Updating a campaign: `src campaigns apply -f foo.campaign.yaml [-namespace alice|myorg] -preview` -> same, except on the backend we know that a campaign in the namespace already exists, so the URL will be like `https://sourcegraph.example.com/users/alice/campaigns/CAMPAIGN_ID/apply?spec=NEW_CAMPAIGN_SPEC_ID`

mrnugget commented on doc/dev/campaigns_design.md:
```diff
@@ -0,0 +1,126 @@
+
+## Principles
+
+- **Declarative API** (not imperative). You declare your intent, such as "lint files in all repositories with a `package.json` file"<!-- TODO(sqs): thorsten had a suggestion to make this quote use non-imperative language but I can't find the comment -->. The campaign figures out how to achieve your desired state. The external state (of repositories, changesets, code hosts, access tokens, etc.) ca ...
```
```suggestion
- **Declarative API** (not imperative). You declare your intent, such as "files in all repositories with a `package.json` file should be linted". The campaign figures out how to achieve your desired state. The external state (of repositories, changesets, code hosts, access tokens, etc.) can change at any time, and temporary errors frequently occur when reading and writing to code hosts. These factors would make an imperative API very cumbersome because each API client would need to handle the complexity of the distributed system.
```

mrnugget commented:
(In addition to the comments above) What's still missing from the API/docs:

1. How to specify existing changesets in a campaign spec.
2. How to preview `applyCampaignSpec`.
3. How to get the current status of the campaignSpec application/reconciliation.

---

Regarding 1.: I think we want something like the following in the campaign spec:

```yaml
# Track changesets that already exist in a repository:
changesetSpecs:
  - repository: github.com/sourcegraph/sourcegraph
    externalID: 1234
  - repository: github.com/sourcegraph/src-cli
   externalID: 6789
```

I'm not sure yet what to call the top-level key here, though. `changesetSpecs` makes sense to me, because I imagine it would be created as a `changesetSpec` and added to the `campaignSpec`. But I can also see how it's confusing next to the `changesetTemplate`.

---

Regarding 2. and 3.: After thinking about this the whole afternoon andreading about [Kubernetes API conventions on "spec and status"](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status) I think the preview functionality and getting the current status of a campaign are highly related.

The preview should tell us:
- which campaign attributes (description, ...) differ between the (current) state and the spec.
- which changesets are in the spec but not in the campaign state. Those will be created & added to campaign or just added to campaign.
- which changesets are in campaign (state) but not in the spec. Those will be removed from the campaign.
- which attributes of which changesets differ between state and spec. Those will be updated.

If we have a concise API representation of this delta, I wonder whether it isn't enough to display enough status information in the UI.

Since the "campaigns controller" (the thing that reconciles spec & state, name is not final, but I think it makes sense) runs — at least from the users perspective — in an infinite loop in the background, we, as API users, can assume that every item in the delta represents an action in progress, as soon as the spec is applied. So if, say, `delta.campaign.attributes.filter(attr => attr.spec != attr.state).length > 0` we know that progress is "Campaign is being updated...".

Ignore the attributes in that last bit of pseudo-code, as I'm not sure yet how we should represent the delta, which is why I'm leaving this comment here, hoping that you all have some ideas :)

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,169 @@ type Mutation {
+    parsedInput: JSONValue!
+
+    # The specs for changesets associated with this campaign.
+    changesetSpecs: [ChangesetSpec!]!
```
I think this needs to be an interface (something like `ChangesetSpecInterface` (names welcome!)) that's implemented by `ChangesetSpec` and `HiddenChangesetSpec`, since we do need to extract the repository for each changeset and make sure that we only show information about the changesetspec if the user has repository permissions for that.

mrnugget commented:
Blocked: https://github.com/sourcegraph/sourcegraph/pull/10921#issuecomment-648845974

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,169 @@ type Mutation {
+    # used in a call to applyCampaign), which can be queried on ChangesetSpec.expiresAt.
+    createChangesetSpec(
+        # The raw changeset spec (as JSON).
+        changesetSpec: String!
```
We also need a schema for this.

mrnugget commented:
Another thing that's still a bit unclear: do we switch to spec-only publishing of changesets? The current API doesn't have `publish*` mutations. Until we know what a workflow looks like that allows users to publish single changeset specs via specification in the YAML, we should probably have these mutations?

sqs commented:
@mrnugget:

> do we switch to spec-only publishing of changesets? The current API doesn't have publish* mutations. Until we know what a workflow looks like that allows users to publish single changeset specs via specification in the YAML, we should probably have these mutations?

Yes, I think we need to switch to spec-only publishing of changesets. We don't want to add an extra place where the user's desired state lives (i.e., by adding a `publishChangeset` mutation and flipping a bit on a changeset when they call that mutation).

If a person wants to publish a changeset, they need to edit the foo.campaign.yaml and add something like:

```
publishChangesets:
  - repo: foo/bar
  - repo: abc
```

or (not sure, will think about, would love feedback)

```
changesetTemplate:
  # ...
  published:
    - repo: foo/bar
    - repo: abc
```

I think this would mean "publish all changesets in repositories whose full name is `foo/bar` or `abc` (for all base branches)".

mrnugget commented:
@sqs okay, that sounds good!

---

I'm trying to keep track of what I'm stilling missing here and what we need to define. Since I found it helpful to summarise my comments above, let me add the list here too, even though it's duplicating some of the comments:

* How do we query the current reconciliation status of a campaign?
* How do we preview campaign updates?
* How do we publish single changesets? How do we publish all of them?
* How do we manually add changesets to a campaign?
* What's the schema for the specs sent up by `src-cli`?
  * What does the JSON/YAML schema for CampaignSpecs look like? i.e. what's the schema of `spec` in `createCampaignSpec(spec: String!)`?
  * What does the JSON/YAML schema for ChangesetSpecs look like? i.e. what's the schema of `spec` in `createChangesetSpec(spec: String!)`?
  * Does the `CampaignSpec` include the `changesetTemplate` or is that expanded client-side and turned into single `ChangesetSpecs`?

cc @eseliger

eseliger commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,169 @@ type Mutation {
+    # The date when this campaign spec was created.
+    createdAt: DateTime
+
+    # The namespace (either a user or organization) of the campaign spec.
```
should we rather prune specs that are in a removed namespace?

mrnugget replied:
Since CampaignSpecs can be attached to campaigns that would mean we'd need to prune campaigns too if a user gets deleted?

eseliger commented:
From [here](https://github.com/sourcegraph/sourcegraph/pull/10921#issuecomment-651786729):
> Yes, I think we need to switch to spec-only publishing of changesets. We don't want to add an extra place where the user's desired state lives (i.e., by adding a publishChangeset mutation and flipping a bit on a changeset when they call that mutation).

That makes me wonder whether we should add a "close flag" to the spec as well. The desired state would be `closed: true`, so if we stick to this model strictly, I think we would need to do it as well. 
However, I'm not super sure we want the UI to basically be read-only completely. Just to gradually roll-out a campaign by publishing one after the other, it feels quite burdensome to always 
pull out an editor, set a boolean flag on one of the changesets and submit it via src-cli. I do see that this is how Kubernetes work and we wanted to loosely inspire our API by that, but I am unsure whether it would make it harder to use. 
When looking into the future, I would imagine there is server-side execution implemented and I can 1. create my manifest in the UI using monaco and 2. pick from a curated list of pre-baked campaigns to make my code a bit healthier. Given that, I would either need to create a file locally to track that, or have that file "tracked" somewhere on the Sourcegraph side and be able to edit it from the UI.
For Kubernetes, however, there is no such thing as "one single source of truth manifest" outside of the k8s api. I can patch the manifest, without knowing the whole original file, I can edit it from another location and have an outdated manifest in my repo and so on. Given that, we can maybe pick this up as well. Update the spec in the DB from the UI by allowing "Publish" (for example) using a button, which then updates the spec. I can in return get the manifest by `src campaign get <id> -o yaml` to update my local (repo) state.

sqs commented:
I removed all impl changes from this PR because they are in https://github.com/sourcegraph/sourcegraph/pull/11675. This PR now contains only docs changes and GraphQL API changes, to make it easier to review.

sqs commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,169 @@ type Mutation {
+    # created. Otherwise, the existing campaign will be updated. The campaign is returned.
+    applyCampaign(
+        # The campaign's namespace (either a user or organization).
+        namespace: ID!
```
Yes, correct. I forgot to remove it here. Done.

sqs commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,169 @@ type Mutation {
+    # between revealing the problem and merging the fixes is as short as possible.
+    id: ID!
+
+    # TODO(sqs): add fields - the main requirements here will be imposed by the Campaign.changesets
+    # field and the CampaignDelta field describing the delta to changesets.
```
That was a good start. Added:

- field `ChangesetSpec.description: ChangesetDescription`
- `union ChangesetDescription = ExistingChangesetReference | GitBranchChangesetDescription`
- types `ExistingChangesetReference` and `GitBranchChangesetDescription`

sqs commented:
@eseliger:

> That makes me wonder whether we should add a "close flag" to the spec as well. The desired state would be `closed: true`, so if we stick to this model strictly, I think we would need to do it as well.

No, I don't think this is necessary. The source of truth for "desire for a changeset to be closed" is the code host, not the campaign spec. (In contrast, the source of truth for "desire for a changeset to be published" is the campaign spec.)

If you "close" a changeset in a campaign, that should translate directly into closing the (eg) pull request on GitHub.

---

So, then, consider this case: You just closed (either on GitHub or via Sourcegraph, which are 100% equivalent) the campaign's pull request in repository `a/b` on branch `foo`---but the campaign spec still says "I want to propose a changeset on repository `a/b` on branch `foo`".

What happens? Well, here's what we want to happen: nothing. The campaign should respect that its changeset was closed.

- If the campaign spec's changesetTemplate changes, the closed PR should not be updated.
- If the campaign spec's command steps change, the branch should not be updated. (It might've been deleted when the PR was closed.)
- If the PR is reopened, then it should start being updated again by the campaign.

---

So, what if we implement a nice, multi-select batch operation UI for changesets, and then you select 500 changesets and click "Close 500 selected changesets"? That is a distributed operation, and it may be tempting to say "at least for implementation purposes, we should model that as though those changeset templates had `closed: false`". Maybe...but the user's intent is a one-time action, not to set a new invariant. Make sure that clicking "Close" on Sourcegraph for a changeset doesn't prevent you from reopening it on the code host (and doesn't undo it if you reopen it on the code host)!

---

("A-ha", you might say, "this is true--but if you close a campaign and want to close associated changesets, that is another distributed operation, and we should (for implementation purposes) represent that as though you set `closed: true` in your campaign spec." I don't think so. Suppose you closed a campaign that was only partially finished being published. You wouldn't want closing it to *finish* publishing all the changesets only to close them immediately. You would just want it to close all the changesets that it had opened. So, this "close changesets when closing campaign" should be handled specially.)

sqs commented:
@eseliger @LawnGnome Check out my updates:

- Files changed:
  - `schema.graphql` types `ChangesetSpec`, `ChangesetDescription`, `ExistingChangesetReference`, `GitBranchChangesetDescription`, `GitCommitDescription`
  - `CampaignSpec.schema.json` for what the YAML campaign file would look like. (The simple example is in the Hello World Campaign doc page linked in this PR description.)
- https://github.com/sourcegraph/sourcegraph/pull/10921#issuecomment-653727228

Feedback appreciated! I am working down the checkboxes at https://github.com/sourcegraph/sourcegraph/pull/11675. (@mrnugget is on vacation.)

mrnugget commented:
> @eseliger @LawnGnome Check out my updates:

Replying here to both changes (`CampaignSpec.schema.json` and the proposed changes in `schema.graphql`): 

If I understand the changes correctly, it would mean that a lot of the fields that are defined in `campaign-spec.yml` are validated _client side_ with the `CampaignSpec.schema.json`, but what will be send up to the server is:

1. `ChangesetSpec`s with full [`ChangesetDescription`](https://github.com/sourcegraph/sourcegraph/pull/10921/files#diff-4529437f5564f5e65e957821a46ff9c5R475-R532)s that are the result of an _evaluated_ `CampaignSpec` and represent the desired state for each changeset in each repository.
2. a `CampaignSpec` that is basically just a copy of the original `campaign-spec.yml` but doesn't contain any additional information, except the namespace.

Correct?

If so, that would be a change from what I've started to implement in https://github.com/sourcegraph/sourcegraph/pull/11675 — a slight change, since I've left that part under-defined on purpose. In the current implementation, the `changesetTemplate` is defined on the `CampaignSpec` (https://github.com/sourcegraph/sourcegraph/pull/11675/files#diff-fa5e1bd8d3af52b49e4ee71fd894780eR1523-R1527) and the `ChangesetSpec` (see [`ChangesetSpecFields`](https://github.com/sourcegraph/sourcegraph/pull/11675/files#diff-fa5e1bd8d3af52b49e4ee71fd894780eR1541-R1567) here)) doesn't contain "title", "body", "commit message" etc. But that implementation also had a lot of questions marks left (manually tracking existing changesets, for example).

I think your proposal is a good step forward and answers multiple of my conceptual/graphql questions in https://github.com/sourcegraph/sourcegraph/pull/11675 but to clarify our shared understanding, I think we really need to answer these two questions in particular:

1. What is the JSON schema for `campaignSpec` in `createCampaignSpec(namespace: ID!, changesetSpecs: [ID!]!, campaignSpec: String!)`?
2. What is the JSON schema for `spec` in `createChangesetSpec(changesetSpec: String!)`?

I think the answer to (1) is the `CampaignSpec.schema.json` that you posted above (because we need the `name` and `description` for the campaign from that, plus we want to keep the original around).

And the answer to (2) is practically the [`ChangesetDescription`](https://github.com/sourcegraph/sourcegraph/pull/10921/files#diff-4529437f5564f5e65e957821a46ff9c5R475-R532).

What do you think?

---

(Implementation note: I'm slightly worried about all the "write" overhead to the database, since `ChangesetsSpecs` are immutable and they all contain a lot of duplicate information. So a user runs `apply -f spec.yml` multiple times, we'd create a ton of duplicate data in the database. It would make the code _a lot simpler and easier to understand_, since the desired state for each changeset is completely contained in the `ChangesetSpec`, but it's also data-heavy)

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -427,6 +367,222 @@ type Mutation {
+    #
+    # Only 1 commit is supported.
+    commits: [GitCommitDescription!]!
+}
```
@sqs @eseliger @LawnGnome Since we said we'll publish not through the UI, but through the spec, shouldn't this description also include something like`Published: Boolean!`?

```suggestion
    published: Boolean!
}
```

mrnugget replied:
@sqs After discussing this with @LawnGnome and @eseliger on a call we think this, the `GitBranchChangesetDescription`, should have `title`, `body` and `published` attributes, since then a `ChangesetSpec` would be a completely evaluated `ChangesetTemplate`, representing the whole "desired state" for a changeset.

sqs replied:
Yep, makes sense.

mrnugget replied:
Just pushed 2dcc0d8 to this branch which adds `title`, `body`, and `published`.

sqs commented:
@mrnugget:

Here is my response based on my thinking. Feel free to challenge my assumptions if you have better ideas.

> If I understand the changes correctly, it would mean that a lot of the fields that are defined in campaign-spec.yml are validated client side with the CampaignSpec.schema.json

Yes, validated and evaluated client-side. The `src` CLI will know how to evaluate the changesetTemplate.

> what will be send up to the server is:

> ChangesetSpecs with full ChangesetDescriptions that are the result of an evaluated CampaignSpec and represent the desired state for each changeset in each repository.

Yes.

> a CampaignSpec that is basically just a copy of the original campaign-spec.yml but doesn't contain any additional information, except the namespace.

Yes, although it could be confusing to allow specifying the namespace in the YAML file and in a CLI flag. See https://github.com/kubernetes/kubernetes/issues/7789 for discussion of how this works for `kubectl` (basically, the `kubectl -n` flag overrides the `metadata.namespace` field; there are some edge cases that don't apply to us). We don't need to allow specifying namespace in the YAML file; we could always just store it separately (always require `src campaign apply -n`, always require the `applyCampaign(namespace: ID!, ...)` namespace parameter, store it in a separate DB column, etc.). Your call, just make sure the precedence is consistent and documented.

> If so, that would be a change from what I've started to implement in #11675 — a slight change, since I've left that part under-defined on purpose. In the current implementation, the changesetTemplate is defined on the CampaignSpec (#11675 (files)) and the ChangesetSpec (see ChangesetSpecFields here)) doesn't contain "title", "body", "commit message" etc. But that implementation also had a lot of questions marks left (manually tracking existing changesets, for example).

Yes. I think it's good to make the ChangesetSpecs comprehensive. This will increase their expressive power. Right now, we don't have a way to have different changeset titles, descriptions, commit messages, etc., among the various changesets in a campaign, but I can certainly see that being useful. (Think of a lint campaign where the changesets say which lint problems were actually found, or where there are different manual steps to revoke different types of tokens leaked in code in addition to just removing them.)

> I think your proposal is a good step forward and answers multiple of my conceptual/graphql questions in #11675 but to clarify our shared understanding, I think we really need to answer these two questions in particular:

> (1) What is the JSON schema for campaignSpec in createCampaignSpec(namespace: ID!, changesetSpecs: [ID!]!, campaignSpec: String!)?
> I think the answer to (1) is the CampaignSpec.schema.json that you posted above (because we need the name and description for the campaign from that, plus we want to keep the original around).

Yes.

> (2) What is the JSON schema for spec in createChangesetSpec(changesetSpec: String!)?
> The ChangesetDescription?

Yes.

mrnugget commented:
I just pushed c346293, which adds a `baseRev` (not `baseRef`! which we already have) field to `GitBranchChangesetDescription`. Strictly speaking we don't need it on the read-end of the API, but since the src-cli needs to send it up and we will have the value at hand I thought I'd add it for completeness sake, consistency and for documentation purposes.

eseliger commented:
__See slack for storybook access.__

After yesterdays call, I've filled the properties of the `CampaignDelta` type and added some storybooks for it, to showcase how it could be used.
See this commit for details:
https://github.com/sourcegraph/sourcegraph/pull/12047/commits/93ede3e7d7bd659338e2c5a4423ce4f7b7bcb27d
Which landed here in this storybook: https://5f078f888555130022c12385-qnepitfatb.chromatic.com/?path=/story/web-campaignupdatediff--campaigndelta-not-empty 
The storybook doesn't reflect a final UI, but rather how useful it could be to have a changeset connection and augment the edges with the update information, so we can still have a single list view with all the changesets in one list. Note that it would require that we always have a Changeset for each spec, even if not applied yet. I think that is doable though and might save us one DB table.

---
The branch also includes changeset state updates, but don't look at them yet, they were based on Adams state diagram. I've created a separate one here: https://docs.google.com/drawings/d/1hicy46Cct9bU3DYdy-uskp1lxQ0KLj5rYdHBYVp7hh8/edit, please take a look there and once we are happy with how it turned out, I'll update the schema and the storybook page here: https://5f078f888555130022c12385-qnepitfatb.chromatic.com/?path=/story/web-changesetnode--visible-changeset

**Work items**
- [ ] Review state diagram and add comments
- [ ] Review Delta API spec (not storybook, it's just for visualization for now)

mrnugget commented:
@LawnGnome @sqs I just discussed Erik's proposed schema (https://github.com/sourcegraph/sourcegraph/commit/95b26aa5e760e93c289f6e89f493e87f5f1f8f4e) and [state diagram](https://docs.google.com/drawings/d/1hicy46Cct9bU3DYdy-uskp1lxQ0KLj5rYdHBYVp7hh8/edit) on a Zoom call.

Can you both also take a look and see if you can spot something that's wrong/could be improved?

The one big thought that just came up in our discussion was: should we create "empty" changesets when we create a changesetSpec (or: do we use a single table for both)? The reasoning behind this question is that the preview-delta-API as proposed by Erik above used `changesets` as the unit of change ("these 5 changesets will be added/closed/updated"). That works well when the changesets already exist (after the user applied a campaignSpec once, for example). But if you preview a `campaignSpec` that has _never been applied_, the delta would need to give show up "these 5 changesets will be created" and since they don't exist yet, they don't have an ID yet, etc. etc.

mrnugget commented:
I thought about this ☝️ some more.

The two biggest question marks we have left in this new flow are:

1. API with which to preview applying of a campaign spec
2. API with which to get the current status of creation/updating/closing/syncing/... of a campaign and its changesets 

Erik's proposal above combines them, by re-using the preview query to get the current status of the campaign. After thinking about this since Friday, I don't think that's a good idea. It mixes two previously unrelated concepts ("status" and "preview") with the hope of getting making things simpler by making them conform to a single concept. The result is complex, though, and would require us to bend over backwards to fit the preview API into the concept of a status API (by requiring us to create changesets for changeset specs that haven't been applied yet, for example).

Instead I think we should take Erik's proposed status API (which, in fact, is more a storybook right now) which would allow API consumers to get the "status" of a campaign by querying its changesets and fully design that.

And then we should design a separate `campaignDelta` API that's based on Quinn's proposed query (`campaignDelta(namespace: ID!, campaignSpec: ID!, ensureCampaign: ID): CampaignDelta!`) and that does _not_ require us to create changesets even though nothing been applied yet. That might mean that we need some nullable fields in that part of the API, but I'm okay with that and I think we're often too scared of the null-boogeyman.

Let's discuss who's going to own the design of these two parts of the API on today's sync, because I think we need more and clearer ownership here :)

sqs commented:
Regarding CampaignDelta (requested by @mrnugget):

- This is a diff, not a "preview/plan of what steps will be taken to achieve a desired state". Because this is a distributed system, any steps produced would be immediately out of date (because things can change at any moment in the external systems that make up the overall state of the system). This is mostly an issue of naming. We will still have a UI that implies "when you apply this campaign spec, these are the changes that will happen". But for the GraphQL API naming, let's avoid usage of terms that could imply that this is a definite preview/plan.
- Should the CampaignDelta compare the current spec *or* the current status (to the new spec)? I don't know.
- We want the UI to be able to show a granular diff of (1) the changes to the campaign itself (which really just means the title, description, and maybe the namespace if that is a CampaignSpec field (see https://github.com/sourcegraph/sourcegraph/pull/10921#issuecomment-656272372)), and (2) the changes to the changesets, including:
  - For added changesets (that are in the new campaign spec and are not in the current one): showing their addition and what will happen (for example, PR will be opened vs. this is just an unpublished changeset),
  - For changed changesets: a granular list of the fields that change, such as PR title, description, diff, commits (each message and author). For v0, we can just indicate *which* fields changed and show both before/after instead of needing to show a super granular diff (such as "commit 3's author name changed").
  - For removed changesets (that are NOT in the new campaign spec and ARE in the current one): showing their removal and what will happen (for example, PR will be closed vs. just an unpublished changeset so no operations need to be performed on the code host).
- I actually think (with certainty 85%) that given the new explicit nature of the changeset specs, which make it really clear what the desired state is, we can postpone the CampaignDelta and it isn't a requirement to graduate from beta. We do still need a way to visually see a single campaign spec and all of its changesets, but we don't need to diff it. The user flow for creating a campaign will be nice and visual; for updating a campaign, they will just need to trust that we'll handle the delta-update correctly.

mrnugget commented:
> * Should the CampaignDelta compare the current spec _or_ the current status (to the new spec)? I don't know.

I think it should compare current status to the new spec. Why? Because we're not 100% declarative, meaning that the state of a changeset can change on a code host (from open to closed, for example) and we won't try to re-open it.

> * I actually think (with certainty 85%) that given the new explicit nature of the changeset specs, which make it really clear what the desired state is, we can postpone the CampaignDelta and it isn't a requirement to graduate from beta. We do still need a way to visually see a single campaign spec and all of its changesets, but we don't need to diff it. The user flow for creating a campaign will be nice and visual; for updating a campaign, they will just need to trust that we'll handle the delta-update correctly.

Can you expand on that?

Do you think a first solution that has "a way to visually see a single campaign spec and all of its changesets" would just be that: the `CampaignSpec.PreviewURL` leads you to a page where the client queries the given campaign spec and displays its changesets?

If that is acceptable as a first solution, I'd prefer to doing just that for now and defer the `CampaignDelta` until we have the rest working.

mrnugget commented:
@sqs Did you have any ideas as to what the "preview" page for a campaign spec would look like?

Now that we said we don't show a delta view there, I suppose it's a list of the CampaignSpec and its ChangesetSpecs. But when talking to @eseliger we realised that there might be different ideas here.

Erik, for example, thought along the lines of "preview page is a variation of the campaign page".

I think it should be a standalone page that looks decidedly different in its layout (so we don't get into the problem that we had with the CampaignDetails view where it was used to display a patchset and patches, a campaign and changesets, a campaign with a patchset and patches and changesets, etc.).

And since we want to eventually implement a delta view, I assume, I don't think we should invest too much into the preview page. On the other hand, though, it's likely the first "campaigns" page that customers will see with the new flow.

sqs commented:
@mrnugget:

> Did you have any ideas as to what the "preview" page for a campaign spec would look like? > Now that we said we don't show a delta view there, I suppose it's a list of the CampaignSpec and its ChangesetSpecs. But when talking to @eseliger we realised that there might be different ideas here.

> Erik, for example, thought along the lines of "preview page is a variation of the campaign page".
> I think it should be a standalone page that looks decidedly different in its layout.

I agree it should look very different in its layout, kind of like how the Amazon checkout flow cuts out a lot of the noisy nav junk. I think it should have no search bar (GlobalNavbar variant `low-profile-with-logo`), no breadcrumbs, just the full campaign spec and changesets with a create-campaign button.

I also think it should show all diffs and data, without the need to selectively expand/collapse.

> And since we want to eventually implement a delta view, I assume, I don't think we should invest too much into the preview page. On the other hand, though, it's likely the first "campaigns" page that customers will see with the new flow.

I think many of the components will be reused between the preview page today, the campaign view page today, and the future delta view. So I don't think this is a dichotomy. The preview page today should look and feel really nice, but I don't think that will take a ton of work. Making it simple (as mentioned above) will go a long way to making it look nice. Another way in which it'll be simpler than in the past is that there will just be changesets, not patches+changesets.
